### ISSUE TITLE
Ember 3.0 Regression: Unable to addObject to an ArrayProxy

### ISSUE THREAD

**Issue by bryanhickerson** [emberjs/ember.js#16451]
In ember 2.18, I am able to use `addObject` to add a new record to an existing collection that was fetched via `queryRecord`. After upgrading to ember 3.0, such an action results in a `The result of a server query (on [model]) is immutable.` error. This also breaks loading additional pages in `ember-infinity` for me (https://github.com/ember-infinity/ember-infinity/issues/270), which uses `pushObjects` and `unshiftObjects` to add subsequent pages to the existing model. I suspect this is related to this change: https://github.com/emberjs/ember.js/pull/16157/files. I understand that immutability is the future direction of ember, but this results in a breaking change that had no deprecation warning. Is it possible to make this less restrictive via deprecation? I'm willing to work on the change, I'm just unsure what the desired approach would be.

**Comment by mmun** [comment]
That error is coming from ember data and has existed for 5+ years. Those record arrays are not intended to be mutable. The PR fixes a bug that was letting you mutate them in some cases (e.g. `.pushObjects()` would let you, but `.replace()` wouldn't).

You can do something like

```js
store.query(...).then(recordArray => recordArray.toArray())
```

to get a copy of the record array that is mutable.

**Comment by bryanhickerson** [comment]
Right, I saw that the error existed in ember data for a long time, but we’ve been able to do that for years in our app without any warning or deprecation. Converting to an array was the approach I took to fix it in my app, but if the model is managed by an add on (ember-infinity), I can’t do that without overriding addon code.

**Comment by mmun** [comment]
[Sorry, deleted my original comment because it was a bit confusing. This one should be clearer.]

In a nutshell what's happening is,

- `RecordArray#pushObjects` is inherited from `ArrayProxy#pushObjects` which is itself inherited  from `MutableArray#pushObjects`
- `MutableArray#pushObjects` calls `this.replace(...)`
- `RecordArray` overrides `replace` throws because it doesn't want you mutating it.

Previously, there was a bug in Ember where `ArrayProxy` was overriding `pushObjects` with its own version that didn't call `this.replace(...)`. This is important and shouldn't change.

So in summary:

- Ember Data has always wanted RecordArrays to be immutable (as the error dates back to 2012!)
- Due to a bug in Ember and poor test coverage in Ember Data, the RecordArrays were in fact mutable if you went through some specific methods
- Ember fixed this bug
- Now, Ember Data is correctly enforcing the RecordArray immutability.

It seems like the right fix here is to make https://github.com/emberjs/data/blob/master/addon/-private/system/record-arrays/record-array.js#L85 a warning rather than revert the bugfix in Ember.

**Comment by bryanhickerson** [comment]
Thanks for the context! I’ll open an ember data pr.
